packages/andromeda_protocol/src/vault.rs

```
@@ -0,0 +1,210 @@
+            Err(_) => Err(ContractError::NotImplemented {
+                msg: Some(String::from("This strategy is not supported by this vault")),
+            }),
+            Ok(addr) => {
+                let msg = wasm_execute(
+                    addr,
+                    &ExecuteMsg::AndrReceive(AndromedaMsg::Receive(Some(to_binary(recipient)?))),
+                    vec![funds],
+                )?;
+                let sub_msg = SubMsg {
+                    id: 1,
+                    msg: CosmosMsg::Wasm(msg),
+                    gas_limit: None,
+                    reply_on: ReplyOn::Error,
+                };
+
+                Ok(sub_msg)
+            }
+        }
+    }
+}
+
+impl ToString for StrategyType {
+    fn to_string(&self) -> String {
+        match self {
+            StrategyType::Anchor => String::from("anchor"),
+            StrategyType::NoStrategy => String::from("none"),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct InstantiateMsg {
+    pub strategies: Vec<YieldStrategy>,
+    pub operators: Option<Vec<String>>,
+}
+
+impl InstantiateMsg {
+    pub fn validate(&self) -> Result<(), ContractError> {
+        for (idx, yield_strategy) in self.strategies.to_vec().iter().enumerate() {
+            if self.strategies.iter().enumerate().any(|(index, strategy)| {
+                index != idx
+                    && (strategy.strategy_type == yield_strategy.strategy_type
+                        || strategy.address == yield_strategy.address)
+            }) {
+                return Err(ContractError::StrategyNotUnique {
+                    strategy: yield_strategy.strategy_type.to_string(),
+                });
+            }
+
+            if yield_strategy.strategy_type == StrategyType::NoStrategy {
+                return Err(ContractError::InvalidStrategy {
+                    strategy: StrategyType::NoStrategy.to_string(),
+                });
+            }
```

nit: Do this before the first check to fail early. Also if `NoStrategy` is invalid, can we just remove that variant?

---

I think it could make sense to have a default strategy if `None` is passed for the `Option<YieldStrategy>` but I can see it going both ways. I think removing `NoStrategy` is fine for now as that feels a bit incompatible with how things are currently handled.